src/utils.c

```diff
@@ -1025,6 +1025,266 @@ ts_relation_size_impl(Oid relid)
+		}
+
+		relation_close(toastRel, AccessShareLock);
+	}
+
+	relation_close(rel, AccessShareLock);
+
+	/* Add up the total size based on the heap size, indexes and toast */
+	relsize.total_size = relsize.heap_size + relsize.index_size + relsize.toast_size;
+
+	return relsize;
+}
+
+TS_FUNCTION_INFO_V1(ts_relation_approximate_size);
+Datum
+ts_relation_approximate_size(PG_FUNCTION_ARGS)
+{
+	Oid relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	RelationSize relsize = { 0 };
+	TupleDesc tupdesc;
+	HeapTuple tuple;
+	Datum values[4] = { 0 };
+	bool nulls[4] = { false };
+
+	/* Build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function returning record called in context "
+						"that cannot accept type record")));
+
+	if (!OidIsValid(relid))
+		PG_RETURN_NULL();
+
+	relsize = ts_relation_approximate_size_impl(relid);
+
+	tupdesc = BlessTupleDesc(tupdesc);
+
+	values[0] = Int64GetDatum(relsize.total_size);
+	values[1] = Int64GetDatum(relsize.heap_size);
+	values[2] = Int64GetDatum(relsize.index_size);
+	values[3] = Int64GetDatum(relsize.toast_size);
+
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+
+	return HeapTupleGetDatum(tuple);
+}
+
+static void
+init_scan_by_hypertable_id(ScanIterator *iterator, int32 hypertable_id)
+{
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(), CHUNK, CHUNK_HYPERTABLE_ID_INDEX);
+	ts_scan_iterator_scan_key_init(iterator,
+								   Anum_chunk_hypertable_id_idx_hypertable_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(hypertable_id));
+}
+
+TS_FUNCTION_INFO_V1(ts_hypertable_approximate_size);
+Datum
+ts_hypertable_approximate_size(PG_FUNCTION_ARGS)
+{
+	Oid relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	RelationSize total_relsize = { 0 };
+	TupleDesc tupdesc;
+	HeapTuple tuple;
+	Datum values[4] = { 0 };
+	bool nulls[4] = { false };
+	Cache *hcache;
+	Hypertable *ht;
+	int32 osm_chunk_id;
+	ScanIterator iterator = ts_scan_iterator_create(CHUNK, RowExclusiveLock, CurrentMemoryContext);
+
+	/* Build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function returning record called in context "
+						"that cannot accept type record")));
+
+	if (!OidIsValid(relid))
+		PG_RETURN_NULL();
+
+	/* go ahead only if this is a hypertable or a CAgg */
+	hcache = ts_hypertable_cache_pin();
+	ht = ts_resolve_hypertable_from_table_or_cagg(hcache, relid, true);
+	if (ht == NULL)
+	{
+		ts_cache_release(hcache);
+		PG_RETURN_NULL();
+	}
+
+	/* get the main hypertable size */
+	total_relsize = ts_relation_approximate_size_impl(relid);
+
+	/* don't think we are interested in OSM chunks */
+	osm_chunk_id = ts_chunk_get_osm_chunk_id(ht->fd.id);
```

Comment:
Isn't simple instead of executing this function just check the `_timescaledb_catalog.chunk.osm_chunk` column inside the iterator?

Reply:
Yeah. Should be better this ways. Done.